cdap-app-fabric/src/main/java/io/cdap/cdap/reporting/ProgramHeartbeatTable.java

```diff
@@ -121,6 +123,32 @@ public void deleteAll() throws IOException {
     return resultRunRecordList;
   }
 
+  /**
+   * Find al running pipelines as of the supplied timestamp, for each of the scanned rows, we maintain the latest {@link
+   * RunRecordDetail} identified by {@link ProgramRunId}, Since there can be more than one RunRecordDetail for the same
+   * runId due to multiple state changes and heart beat messages.
+   *
+   * @param namespaces set of namespaces
+   * @return collection of {@link RunRecordDetail}
+   */
+  public Collection<RunRecordDetail> findRunningAtTimestamp(long runningOnTimestamp,
+                                                            Set<String> namespaces) throws IOException {
+    List<RunRecordDetail> resultRunRecordList = new ArrayList<>();
+    for (String namespace : namespaces) {
+      //Scan from the beggining of time until the desired timestamp.
+      List<Field<?>> startRowKey = getScanKey(namespace, 0);
+      List<Field<?>> endRowKey = getScanKey(namespace, runningOnTimestamp);
+      performScanAddToList(startRowKey, endRowKey, resultRunRecordList);
+    }
+
+    // Keep only records that are in RUNNING status
+    resultRunRecordList = resultRunRecordList.stream()
+      .filter(rrd -> rrd.getStatus() == ProgramRunStatus.RUNNING)
```

Comment:
It's better to push this predicate into the `performScanAddToList` method to reduce the size of the in memory result list.

Reply:
I have implemented this change.